rclpy/rclpy/node.py

```diff
@@ -956,6 +958,66 @@ def _set_parameters_atomically_common(
 
         return result
 
+    def list_parameters(
+        self,
+        prefixes: List[str],
+        depth: int
+    ) -> ListParametersResult:
+        """
+        Get a list of parameter names and their prefixes.
+
+        :param prefixes: A list of prefixes to filter the parameter names.
+            Only the parameter names that start with any of the prefixes are returned.
+            If empty, all parameter names are returned.
+        :param depth: The depth of nested parameter names to include.
+        :return: The list of parameter names and their prefixes.
+        :raises: TypeError if the type of the passed argument is not expected types.
+        :raises: ValueError if depth is negative integer.
+        """
+        if not isinstance(prefixes, list):
+            raise TypeError('The prefixes argument must be a list')
+        if not all(isinstance(prefix, str) for prefix in prefixes):
+            raise TypeError('All prefixes must be instances of type str')
+        if not isinstance(depth, int):
+            raise TypeError('The depth must be instance of type int')
+        if depth < 0:
+            raise ValueError('The depth must be greater than or equal to zero')
+
+        result = ListParametersResult()
+
+        separator_less_than_depth: Callable[[str], bool] = \
+            lambda str: str.count(PARAMETER_SEPARATOR_STRING) < depth
+
+        recursive: bool = \
+            (len(prefixes) == 0) and (depth == ListParameters.Request.DEPTH_RECURSIVE)
+
+        for param_name in self._parameters.keys():
+            if not recursive:
+                get_all: bool = (len(prefixes) == 0) and (separator_less_than_depth(param_name))
+                if not get_all:
+                    prefix_matches = any(
+                        param_name == prefix or
+                        (
+                            param_name.startswith(prefix+PARAMETER_SEPARATOR_STRING) and
+                            (
+                                depth == ListParameters.Request.DEPTH_RECURSIVE or
+                                separator_less_than_depth(param_name[len(prefix)+1:])
```

Comment:
The `separator_less_than_depth(param_name[len(prefix)+1:])` is the only different line compared to the [one in rclcpp](https://github.com/ros2/rclcpp/blob/ba175922d3008704c67b10fe82a2f7c0cbee2be4/rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp#L1062-L1063).

I believe [the test case below](https://github.com/ros2/rclpy/pull/1124/files#diff-e2598f8cd62aef1be16dd72039bf6a457f42521b41f54dedd3831b2b1ee01425R1275-R1296) should be available since the depth is relative to the given prefixes. However, if this is present in `separator_less_than_depth(param_name[len(prefix):])`, it fails. (it passes if the depth is 2 or more)
If this understanding is correct, the one in rclcpp also should be fixed.

Reply:
That is a really good call.  You are right that that test should work.

Do you mind submitting a pull request to `rclcpp` that fixes the bug and adds a test for it?

Reply:
:ok_hand: Please check [this](https://github.com/ros2/rclcpp/pull/2300).

Reply:
Thank you!